For #435 - add dynamic and nested claim reasons

princefishthrower opened the pull request:
In the same spirit as return reasons - added a new model, `claim-reason.ts` and linked it up to the existing `claim-item.ts` model while deleting the old enum. Also wrote the service and API layers for these claim reasons (both essentially copied from the return reasons and changing all applicable variables and texts.)

princefishthrower commented:
@srindom - any comment or update here? Thanks!

srindom commented:
Really great work you've done! Will follow up on your questions in #435:

Yes it would be great to support TS throughout and it will be relevant to begin gradually replacing elements over time; I think we will need to discuss an approach before beginning the process though. If you have ideas I'd be happy to open a discussion topic on it :)

Regarding testing we ask that you provide both unit tests and integration tests in your PRs. 

You can find the unit tests in `packages/*/src/services/__tests__` and `packages/*/src/api/routes/*/__tests__` and integration tests in `integration-tests/*/__tests__`

To run unit tests you can do:
```
cd packages/medusa
yarn test
```

Integration tests are a bit trickier:
1. First make sure you have compiled all files. Do this with `yarn bootstrap` in the monorepo root.
2. `cd integration-tests/api`
3. `medusa-dev --scan-once` <-- if you don't have the dev cli installed `yarn global add medusa-dev-cli`
4. `yarn test`

**Quick note on `medusa-dev`**
The Medusa core package and the plugins we provide are all distributed as NPM packages and it would be quite annoying to have to publish to NPM every time you wanted to test a change you have made. To make it easy to get your local changes into a dummy project we therefore provide a dev cli. The first time you use it you will have to specify the path to your monorepo root; from there on the tool will copy all medusa packages from your local monorepo into a given project's node_modules. The figure below illustrates how local development differs from regular usage of Medusa packages:

![image](https://user-images.githubusercontent.com/7554214/135821356-6be6b9de-aa3a-46c6-a532-4161d61187e1.png)
 
 Let me know if this makes sense :)

srindom commented:
Would appreciate if you can add unit and integration tests. You may find this file helpful: https://github.com/medusajs/medusa/blob/master/integration-tests/api/__tests__/admin/return-reason.js

Also the way claims are handled will likely need to be updated. Example of current integration test that will fail->  https://github.com/medusajs/medusa/blob/27ebe2aaf9a15b192ca88f5d7feb20e06e8b2932/integration-tests/api/__tests__/admin/order.js#L339

princefishthrower commented:
@srindom - I am stuck on a train all day tomorrow, so I guarantee there will be an update to this pull request by the end of the day tomorrow! Passing tests and all :) 👍

princefishthrower commented:
@srindom  - I've just added the unit and integration tests. I changed "production_failure" to "any_custom_claim_reason_string" to signify that it can be any reason really, and I've removed the enum value check from the unit test. For integration tests, I've added `claim-reason.js` to both `admin` and `store`, adapting it from `return-reason.js` But actually running the unit tests on my machine, I had some issues. Starting with unit tests, I'm unable to issue `yarn install` in `packages/medusa`, I get:

```
ode:events:368
      throw er; // Unhandled 'error' event
      ^

Error: EISDIR: illegal operation on a directory, read
Emitted 'error' event on ReadStream instance at:
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -21,
  code: 'EISDIR',
  syscall: 'read'
}
```

Then, for integration tests, when I try `yarn bootstrap`, I get:

```
yarn run v1.22.10
$ lerna bootstrap
lerna notice cli v4.0.0
lerna info versioning independent
lerna info Bootstrapping 32 packages
lerna info Symlinking packages and binaries
lerna info lifecycle medusa-telemetry@0.0.5~postinstall: medusa-telemetry@0.0.5
lerna WARN lifecycle The node binary used for scripts is /var/folders/j1/ydw6655510z_5szr12ptrhn40000gn/T/yarn--1634380751305-0.6722329035750425/node but npm is using /Users/chris/.nvm/versions/node/v16.10.0/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> medusa-telemetry@0.0.5 postinstall /Users/chris/opensource/medusa/packages/medusa-telemetry
> node dist/postinstall.js || true

node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '/Users/chris/opensource/medusa/packages/medusa-telemetry/dist/postinstall.js'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
lerna info lifecycle create-medusa-app@0.0.3~prepare: create-medusa-app@0.0.3
lerna info lifecycle medusa-core-utils@1.1.22~prepare: medusa-core-utils@1.1.22
lerna info lifecycle medusa-plugin-discount-generator@1.1.15~prepare: medusa-plugin-discount-generator@1.1.15
lerna info lifecycle medusa-plugin-ip-lookup@1.1.15~prepare: medusa-plugin-ip-lookup@1.1.15
lerna info lifecycle medusa-dev-cli@0.0.20~prepare: medusa-dev-cli@0.0.20
lerna info lifecycle medusa-telemetry@0.0.5~prepare: medusa-telemetry@0.0.5

> create-medusa-app@0.0.3 prepare /Users/chris/opensource/medusa/packages/create-medusa-app
> yarn build

$ microbundle -i src/index.ts --no-pkg-main --target=node -f=cjs --sourcemap=false --compress --alias worker_threads=@ascorbic/worker-threads-shim
rpt2: options error TS5069: Option 'declarationDir' cannot be specified without specifying option 'declaration' or option 'composite'.
Build "createMedusaApp" to dist:
       133 kB: index.js.gz
      92.5 kB: index.js.br

> medusa-core-utils@1.1.22 prepare /Users/chris/opensource/medusa/packages/medusa-core-utils
> cross-env NODE_ENV=production npm run build


> medusa-core-utils@1.1.22 build
> babel src --out-dir dist/ --ignore **/__tests__

Successfully compiled 11 files with Babel (1243ms).

> medusa-plugin-discount-generator@1.1.15 prepare /Users/chris/opensource/medusa/packages/medusa-plugin-discount-generator
> cross-env NODE_ENV=production npm run build


> medusa-plugin-discount-generator@1.1.15 build
> babel src -d .

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
Successfully compiled 2 files with Babel (818ms).
lerna info lifecycle medusa-fulfillment-manual@1.1.22~prepare: medusa-fulfillment-manual@1.1.22
lerna info lifecycle medusa-fulfillment-webshipper@1.1.26~prepare: medusa-fulfillment-webshipper@1.1.26
lerna info lifecycle medusa-interfaces@1.1.23~prepare: medusa-interfaces@1.1.23
lerna info lifecycle medusa-test-utils@1.1.25~prepare: medusa-test-utils@1.1.25

> medusa-dev-cli@0.0.20 prepare /Users/chris/opensource/medusa/packages/medusa-dev-cli
> cross-env NODE_ENV=production npm run build


> medusa-dev-cli@0.0.20 build
> babel src --out-dir dist --ignore "**/__tests__"

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

Why you should do it regularly:
https://github.com/browserslist/browserslist#browsers-data-updating
Error: [BABEL]: Cannot find module '@babel/core'
Require stack:
- /Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/parameter/parameterVisitor.js
- /Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/plugin.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/module-types.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/configuration.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/index.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/index.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/lib/babel/options.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/lib/babel/index.js
- /Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/bin/babel.js (While processing: /Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/plugin.js)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/parameter/parameterVisitor.js:8:13)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/parameter/parameterVisitor.js',
    '/Users/chris/opensource/medusa/packages/babel-preset-medusa-package/node_modules/babel-plugin-transform-typescript-metadata/lib/plugin.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/module-types.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/configuration.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/config/files/index.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/core/lib/index.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/lib/babel/options.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/lib/babel/index.js',
    '/Users/chris/opensource/medusa/packages/medusa-dev-cli/node_modules/@babel/cli/bin/babel.js'
  ]
}
lerna info lifecycle medusa-dev-cli@0.0.20~prepare: Failed to exec prepare script

> medusa-plugin-ip-lookup@1.1.15 prepare /Users/chris/opensource/medusa/packages/medusa-plugin-ip-lookup
> cross-env NODE_ENV=production npm run build

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
🚀 chris@Macbook-Pro [/Users/chris/opensource/medusa] %
> medusa-plugin-ip-lookup@1.1.15 build
> babel src --out-dir . --ignore **/__tests__

Successfully compiled 2 files with Babel (643ms).
```

and the terminal appears to hang and never closes... unless that message `Successfully compiled 2 files with Babel (643ms).` actually "is" a success message, despite the terminal output? Would be great to get some feedback or troubleshooting here so I can actually run the tests.
